File: frontend/src/api/alerts.ts
import { getClientId } from '../utils/clientId'

const BASE = (import.meta.env.VITE_API_URL ?? '') + '/api'

export type AlertType = 'pct_move' | 'abs_price'
export type AlertDirection = 'above' | 'below'
export type DeliveryMode = 'in_app' | 'in_app_push'

export interface AlertRule {
  id: string
  yf_symbol: string
  symbol: string
  name: string
  portfolio: string
  type: AlertType
  direction: AlertDirection
  reference_value: number
  threshold_value: number
  enabled: boolean
  triggered: boolean
  created_at: number
}

export interface AlertNotification {
  id: string
  rule_id: string
  yf_symbol: string
  symbol: string
  name: string
  portfolio: string
  message: string
  triggered_at: number
  read: boolean
}

export interface AlertSettings {
  delivery_mode: DeliveryMode
}

async function req<T>(path: string, init?: RequestInit): Promise<T> {
  const sep = path.includes('?') ? '&' : '?'
  const url = `${BASE}${path}${sep}client_id=${encodeURIComponent(getClientId())}`
  const res = await fetch(url, {
    ...init,
    headers: { 'Content-Type': 'application/json', ...(init?.headers ?? {}) },
  })
  if (!res.ok) {
    const text = await res.text().catch(() => '')
    throw new Error(`API ${res.status}: ${text}`)
  }
  return res.json() as Promise<T>
}

export function fetchAlertRules(): Promise<{ rules: AlertRule[] }> {
  return req('/alerts/rules')
}

export function createAlertRule(body: {
  yf_symbol: string
  symbol: string
  name?: string
  portfolio?: string
  type: AlertType
  direction: AlertDirection
  reference_value?: number
  threshold_value: number
}): Promise<{ rule: AlertRule }> {
  return req('/alerts/rules', { method: 'POST', body: JSON.stringify(body) })
}

export function updateAlertRule(
  id: string,
  patch: Partial<{
    type: AlertType
    direction: AlertDirection
    reference_value: number
    threshold_value: number
    enabled: boolean
    rearm: boolean
  }>,
): Promise<{ rule: AlertRule }> {
  return req(`/alerts/rules/${id}`, { method: 'PATCH', body: JSON.stringify(patch) })
}

export function deleteAlertRule(id: string): Promise<{ ok: true }> {
  return req(`/alerts/rules/${id}`, { method: 'DELETE' })
}

export function fetchAlertNotifications(): Promise<{ notifications: AlertNotification[] }> {
  return req('/alerts/notifications')
}

export function readAlertNotification(id: string): Promise<{ notification: AlertNotification }> {
  return req(`/alerts/notifications/${id}/read`, { method: 'POST' })
}

export function dismissAlertNotification(id: string): Promise<{ ok: true }> {
  return req(`/alerts/notifications/${id}/dismiss`, { method: 'POST' })
}

export function fetchAlertSettings(): Promise<AlertSettings> {
  return req('/alerts/settings')
}

export function updateAlertSettings(delivery_mode: DeliveryMode): Promise<AlertSettings> {
  return req('/alerts/settings', { method: 'POST', body: JSON.stringify({ delivery_mode }) })
}

export function fetchVapidPublicKey(): Promise<{ key: string }> {
  return req('/alerts/vapid-public-key')
}

export function subscribePush(subscription: PushSubscriptionJSON): Promise<{ subscription: unknown }> {
  return req('/alerts/subscribe', { method: 'POST', body: JSON.stringify(subscription) })
}

export function unsubscribePush(endpoint: string): Promise<{ ok: true }> {
  return req('/alerts/unsubscribe', { method: 'POST', body: JSON.stringify({ endpoint }) })
}

File: backend/routers/alerts.py
"""
Price alerts API — rules (create/edit/delete/rearm), notifications
(list/read/dismiss), the delivery-mode setting, and Web Push subscription
management.

Every endpoint resolves the caller via get_current_user (backend/auth_deps.py)
from their session JWT — as of 2026-08-13 this replaced the old
client-supplied `client_id` query param (blind trust was fine for a
throwaway anonymous UUID, not fine now that real accounts exist). All state
is still partitioned per identity in backend/alerts_store.py — its
`client_id` parameter name is unchanged, it's just fed a verified email now
instead of a self-reported UUID — so different people sharing this
deployment never see each other's rules/notifications, and a push
notification only ever reaches the device whose rule fired.

Evaluation itself happens in backend/alerts_engine.py, called from
price_refresh.py's background loop — this router is pure CRUD over
backend/alerts_store.py.
"""

from __future__ import annotations

import os

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from backend import alerts_store as store
from backend.auth_deps import get_current_user

router = APIRouter()


# ── Rules ─────────────────────────────────────────────────────────────────

class CreateRuleRequest(BaseModel):
    yf_symbol:       str
    symbol:          str
    name:            str = ""
    portfolio:       str = ""
    type:            str            # 'pct_move' | 'abs_price'
    direction:       str            # 'above' | 'below'
    reference_value: float = 0
    threshold_value: float


class UpdateRuleRequest(BaseModel):
    type:            str | None = None
    direction:       str | None = None
    reference_value: float | None = None
    threshold_value: float | None = None
    enabled:         bool | None = None
    rearm:           bool | None = None   # if true, resets triggered=False


@router.get("/api/alerts/rules")
def list_rules(client_id: str = Depends(get_current_user)):
    return {"rules": store.get_rules(client_id)}


@router.post("/api/alerts/rules")
def create_rule(body: CreateRuleRequest, client_id: str = Depends(get_current_user)):
    if body.type not in ("pct_move", "abs_price"):
        raise HTTPException(400, "type must be 'pct_move' or 'abs_price'")
    if body.direction not in ("above", "below"):
        raise HTTPException(400, "direction must be 'above' or 'below'")
    rule = store.add_rule(client_id, body.model_dump())
    return {"rule": rule}


@router.patch("/api/alerts/rules/{rule_id}")
def edit_rule(rule_id: str, body: UpdateRuleRequest, client_id: str = Depends(get_current_user)):
    patch = body.model_dump(exclude_unset=True, exclude={"rearm"})
    if body.rearm:
        rule = store.rearm_rule(client_id, rule_id, patch.get("reference_value"))
    elif patch:
        rule = store.update_rule(client_id, rule_id, patch)
    else:
        rule = next((r for r in store.get_rules(client_id) if r["id"] == rule_id), None)
    if rule is None:
        raise HTTPException(404, "rule not found")
    return {"rule": rule}


@router.delete("/api/alerts/rules/{rule_id}")
def remove_rule(rule_id: str, client_id: str = Depends(get_current_user)):
    if not store.delete_rule(client_id, rule_id):
        raise HTTPException(404, "rule not found")
    return {"ok": True}


# ── Notifications ─────────────────────────────────────────────────────────

@router.get("/api/alerts/notifications")
def list_notifications(client_id: str = Depends(get_current_user)):
    return {"notifications": store.get_notifications(client_id)}


@router.post("/api/alerts/notifications/{notification_id}/read")
def read_notification(notification_id: str, client_id: str = Depends(get_current_user)):
    notification = store.mark_notification_read(client_id, notification_id)
    if notification is None:
        raise HTTPException(404, "notification not found")
    return {"notification": notification}


@router.post("/api/alerts/notifications/{notification_id}/dismiss")
def dismiss_notification(notification_id: str, client_id: str = Depends(get_current_user)):
    if not store.dismiss_notification(client_id, notification_id):
        raise HTTPException(404, "notification not found")
    return {"ok": True}


# ── Settings ──────────────────────────────────────────────────────────────

class SettingsRequest(BaseModel):
    delivery_mode: str   # 'in_app' | 'in_app_push'


@router.get("/api/alerts/settings")
def get_settings(client_id: str = Depends(get_current_user)):
    return store.get_settings(client_id)


@router.post("/api/alerts/settings")
def update_settings(body: SettingsRequest, client_id: str = Depends(get_current_user)):
    if body.delivery_mode not in ("in_app", "in_app_push"):
        raise HTTPException(400, "delivery_mode must be 'in_app' or 'in_app_push'")
    return store.update_settings(client_id, body.model_dump())


# ── Push subscriptions ───────────────────────────────────────────────────

class PushKeys(BaseModel):
    p256dh: str
    auth:   str


class SubscribeRequest(BaseModel):
    endpoint: str
    keys:     PushKeys


class UnsubscribeRequest(BaseModel):
    endpoint: str


@router.get("/api/alerts/vapid-public-key")
def vapid_public_key():
    key = os.environ.get("VAPID_PUBLIC_KEY", "").strip()
    if not key:
        raise HTTPException(503, "push notifications not configured on this server")
    return {"key": key}


@router.post("/api/alerts/subscribe")
def subscribe(body: SubscribeRequest, client_id: str = Depends(get_current_user)):
    sub = store.add_subscription(client_id, body.model_dump())
    return {"subscription": sub}


@router.post("/api/alerts/unsubscribe")
def unsubscribe(body: UnsubscribeRequest, client_id: str = Depends(get_current_user)):
    store.remove_subscription(client_id, body.endpoint)
    return {"ok": True}
